jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+Remark **for chas:** currently when we schedule a trigger everything hangs, waiting for the appropriate time to arrive. This stalls camera and sync analysis. Don't you think we should really figure out a way of doing this asynchronously, without blocking (or come to a definite conclusion that blocking is the only reliable enough way to do it on the Pi).
+
+**For me: where do the stages come in?** cli.py sets up the usb/serial link, but where is it used, by whom, and when? Should be clearer after a refactor. Either we need to conditionally interact with the stages, or we do it on a more distributed basis, in response to events. That's more how I do it in my code, and I like that approach.
+
+determine_reference_period.py
+
+**For me and chas:** here and elsewhere, I'm happy to help refactor into a class myself. This module definitely needs some sort of refactoring (see "WTF"/"Ugh" comments).
+
+**For chas**: padding flag seems to be used fairly inconsistently (would affect the if test in establish_indices), and when False they *still* seem to include extra ref frames, just fewer of them. Weird.... In fact, do you ever *not *use it, or should we just remove the parameter?
+
+**For chas**: **IMPORTANT:** there is a structural issue here where the period-determination code processes one big block of frame buffers all together, once. That is not a good solution when running live - it will be much better to make a new attempt to determine a period after each individual new frame is received, rather than doing a massive chunk of processing in the loop in establish_indices. I really think this needs a refactor. Otherwise, apart from anything else, we will end up acquiring significantly more frames than we need to before we can establish a period. There are also scenarios where this approach will fail, where it could succeed if it was allowed just to keep trying indefinitely, processing a rolling buffer of recent frames. (Minor related observation: once that is changed, and the loop is removed from this function, the "didn't find a period" error will not be "critical" when running live)
+
+prospective_optical_gating.py
+
+[No comments beyond the inline ones in the source code]
+
+emulate.py
+
+I haven't really read through this in detail yet, but I do remember that I didn't find it particularly helpful as an example to follow when I was trying to do my own offline analysis using this codebase. I need to go back and remind myself what the difficulties were that I encountered...
+
+stage_control_functions.py
+
+Looks pretty well commented and structured, but I would implement this as a class, so that other (sub-)classes could be substituted to control other stage types, etc. I will give this some more thought. **For me**: how is this used? Which object is the most sensible one to be in charge of all this?
```

Comment:
Agreed about the class thing. This code has also been ported to the playspim repo. Daniel did one set of stages, I copied these over. Daniel did a class, I stuck with a module. The two sets of codes are almost identical but his class codes make it just a little bit simpler. It is on my todo list to classify those playspim codes. As said above though, I think that the stages should possibly be flat out removed?

Reply:
Stages now removed from this codebase @jmtayloruk